source/iotagent-architecture.rst

```diff
@@ -0,0 +1,543 @@
+no device from one tenant can send messages to any device from other tenant.
+
+The mechanism through which context separation is implemented highly depends on
+which protocol is used. A thorough analysis should be performed to properly
+implement this feature.
+
+IoT agent information and management
+====================================
+
+An IoT agent should expose all the necessary information to use it properly. It
+should expose:
+
+- **Device template**: an IoT agent should publish which is the data model it
+  accepts for a valid device. This should be done by publishing a new device
+  template to other dojot services. There should be a mechanism so that
+  different instances of the same IoT agent publishes the same device template
+  (including any template IDs). If the device template is updated in a newer
+  version of an IoT agent, the device template ID should change. The messages
+  used to publish its device template is detailed in `Device Template message`_
+  section.
+- **Management APIs**: an IoT agent should be manageable and should expose its
+  APIs to to that. The minimum set of management APIs that an IoT agent should
+  offer are:
+
+  - *Logging*: there should be a way to changing the log level of an IoT agent;
+  - *Statistics*: an IoT agent may expose an API to let a user or application
+    retrieve statistical information about its execution. An administrator
+    might want to trigger on or off the generation of a particular statistical
+    variable, such as processing time.
+
+An IoT agent should also be able to gather statistics information related to
+its execution. Furthermore, it should let an administrator to set quotas on
+those measured quantities. These quantities might include, but not limited to:
+
+- transmission statistics
+
+  - number of received device messages from device (total, per device, per
+    tenant)
+  - number of published device messages to dojot (total, per device, per
+    tenant)
+  - number of messages sent to devices (total, per device, per tenant)
+  - [optional] time taken between receiving a message from a physical device
+    and publishing it (total - mean, per device - mean, per tenant - mean)
+
+- IoT agent service health check
+  - system statistics (memory, disk, etc.) used by the service
+
+Many other values might be gathered. The list above is the minimum list that an
+IoT agent is expected to expose to other services. Particularly for health
+check, there is a document detailing how expose it.
+
+IoT agent operation
+===================
+
+The main purpose of an IoT agent is to publish data from a particular device to
+other dojot services. Its operation is two fold: receive and process messages
+related to device management from other services as well as receive messages
+from the devices themselves (or their representative elements) and publish
+these data to other services. This is shown in the figure below.
+
+<< a figure should be here >>
```

Comment:
Where is this figure?

Reply:
:notes:  Sooooomewhere over the rainbow way up high... :notes:

Reply:
I'll add later.